classes/submission/SubmissionFileDAO.inc.php

```diff
@@ -781,6 +783,89 @@ public function getWorkflowStageId($submissionFile) {
 		}
 	}
 
+	/**
+	 * Get the file stage ids that a user can access based on their
+	 * stage assignments
+	 *
+	 * This does not return file stages for ROLE_ID_REVIEWER or ROLE_ID_READER.
+	 * These roles are not granted stage assignments and this method should not
+	 * be used for these roles.
+	 *
+	 * This method does not define access to review attachments or discussion
+	 * files. Access to these files are not determined by stage assignment.
+	 *
+	 * In some cases it may be necessary to apply additional restrictions. For example,
+	 * authors are granted write access to submission files or revisions only when other
+	 * conditions are met. This method does not return those as allowed file
+	 * stages for author assignments.
+	 *
+	 * @param array $stageAssignments The stage assignments of this user.
+	 *   Each key is a workflow stage and value is an array of assigned roles
+	 * @param int $action Read or write to file stages. One of SUBMISSION_FILE_ACCESS_
+	 * @return array List of file stages (SUBMISSION_FILE_*)
+	 */
+	public function getAssignedFileStageIds($stageAssignments, $action) {
```

Comment:
This feels more like a service class function than a DAO function to me (or in much older -- and very deprecated -- terms, an Action class). I'm OK to leave it here but wouldn't want this kind of code to get too entrenched in DAOs without broader consideration.

Reply:
Yep! It is a service class method in `master`. We just don't have `SubmissionFileService` in the stable branches.